src/hotspot/share/runtime/atomic.hpp

```diff
@@ -1232,4 +1251,492 @@ inline T Atomic::XchgUsingCmpxchg<byte_size>::operator()(T volatile* dest,
+  } else {
+    static_assert(dependent_always_false<T>, "Invalid atomic value type");
+  }
+}
+
+// pre-C++17 selection of AtomicValue implementation from T.
+// Uses IntegerValueSelector, ValueSelectorImpl, and ValueSelector.
+template<typename T>
+class Atomic::IntegerValueSelector<T, std::enable_if_t<(sizeof(T) >= sizeof(int))>> {
+public:
+  using type = IntegerValue<T>;
+};
+
+template<typename T>
+class Atomic::IntegerValueSelector<T, std::enable_if_t<(sizeof(T) == 1)>> {
+public:
+  using type = ByteValue<T>;
+};
+
+template<typename T>
+class Atomic::ValueSelectorImpl<
+  T,
+  std::enable_if_t<std::is_integral<T>::value>> {
+public:
+  using type = typename IntegerValueSelector<T>::type;
+};
+
+template<typename T>
+class Atomic::ValueSelectorImpl<T*> {
+public:
+  using type = PointerValue<T*>;
+};
+
+template<typename T>
+class Atomic::ValueSelectorImpl<
+  T,
+  std::enable_if_t<PrimitiveConversions::Translate<T>::value>>
+{
+public:
+  using type = TranslatingValue<T>;
+};
+
+template<typename T>
+class Atomic::ValueSelector {
+  static_assert(std::is_same<T, std::remove_cv_t<T>>::value,
+                "Value type must not be cv-qualified");
+
+public:
+  using type = typename ValueSelectorImpl<T>::type;
+};
+// End of pre-C++17 selection of AtomicValue implementation from T.
+
+template<typename T>
+class Atomic::ValueAccess {
+  T volatile _value;
+
+protected:
+  explicit ValueAccess(T value) : _value(value) {}
+  ~ValueAccess() = default;
+
+public:
+  NONCOPYABLE(ValueAccess);
+
+  // Used internally, not part of final derived public API.
+  T volatile* value_ptr() { return &_value; }
+  T const volatile* value_ptr() const { return &_value; }
+
+  // These are mostly for assembly language access to the value.
+  static int value_offset_in_bytes() { return offset_of(ValueAccess, _value); }
+  static int value_size_in_bytes() { return sizeof(_value); }
+
+  T load_relaxed() const { return Atomic::load(value_ptr()); }
+  T load_acquire() const { return Atomic::load_acquire(value_ptr()); }
+
+  void relaxed_store(T value) { Atomic::store(value_ptr(), value); }
+  void release_store(T value) { Atomic::release_store(value_ptr(), value); }
+  void release_store_fence(T value) { Atomic::release_store_fence(value_ptr(), value); }
+
+  T cmpxchg(T compare_value, T exchange_value,
+            atomic_memory_order order = memory_order_conservative) {
+    return Atomic::cmpxchg(value_ptr(), compare_value, exchange_value);
+  }
+};
+
+template<typename T>
+class Atomic::ValueSupportsXchg : public ValueAccess<T> {
+  using Base = ValueAccess<T>;
+
+protected:
+  explicit ValueSupportsXchg(T value) : Base(value) {}
+  ~ValueSupportsXchg() = default;
+
+public:
+  using Base::value_ptr;
+
+  using Base::value_offset_in_bytes;
+  using Base::value_size_in_bytes;
+  using Base::load_relaxed;
+  using Base::load_acquire;
+  using Base::relaxed_store;
+  using Base::release_store;
+  using Base::release_store_fence;
+  using Base::cmpxchg;
```

Comment:
I waffled back and forth about these long lists of using declarations. Instead maybe macroize?